src/System.Reflection.Extensions/src/System/Reflection/CustomAttributeSearcher.cs

```diff
@@ -0,0 +1,202 @@
+                    //
+                    // Since "typeFilterKnownToBeSealed" is only used to enable an optimization, it's always safe to leave it "false".
+                    //
+                    // Because the Project N toolchain removes any custom attribute that refuses to opt into metadata so at this point,
+                    // we could simply return an empty enumeration and "be correct." However, the code paths following this already do that naturally.
+                    // (i.e. the "passFilter" will never return true, thus we will never attempt to query the custom attribute type for its
+                    // own AttributeUsage custom attribute.) If the toolchain behavior changes in the future, it's preferable that 
+                    // this shows up as new MissingMetadataExceptions rather than incorrect results from the api so we will not put 
+                    // in an explicit return here. 
+                }
+            }
+
+            Func<Type, bool> passesFilter;
+            if (optionalAttributeTypeFilter == null)
+            {
+                passesFilter =
+                    delegate (Type actualType)
+                    {
+                        return true;
+                    };
+            }
+            else
+            {
+                passesFilter =
+                    delegate (Type actualType)
+                    {
+                        if (optionalAttributeTypeFilter.Equals(actualType))
+                            return true;
+                        if (typeFilterKnownToBeSealed)
+                            return false;
+                        return actualType.GetTypeInfo().IsSubclassOf(optionalAttributeTypeFilter);
+                    };
+            }
+
+            return GetMatchingCustomAttributesIterator(element, passesFilter, inherit);
+        }
+
+        //
+        // Subclasses should override this to compute the "parent" of the element for the purpose of finding "inherited" custom attributes.
+        // Return null if no parent.
+        //
+        public virtual E GetParent(E e)
+        {
+            return null;
+        }
+
+
+        //
+        // Main iterator.
+        //
+        private IEnumerable<CustomAttributeData> GetMatchingCustomAttributesIterator(E element, Func<Type, bool> rawPassesFilter, bool inherit)
+        {
+            Func<Type, bool> passesFilter =
+                delegate (Type attributeType)
+                {
+                    // Windows prohibits instantiating WinRT custom attributes. Filter them from the search as the desktop CLR does.
+                    TypeAttributes typeAttributes = attributeType.GetTypeInfo().Attributes;
+                    if (0 != (typeAttributes & TypeAttributes.WindowsRuntime))
+                        return false;
+                    return rawPassesFilter(attributeType);
+                };
```

Comment:
We don't need to construct a new delegate here, which also involves a closure.  Just move the body (with a few tweaks) into the for loop below.